test/TritonGPU/loop-pipeline-hopper.mlir

```diff
@@ -521,8 +521,8 @@ module attributes {"triton_gpu.compute-capability" = 90 : i32, "triton_gpu.num-c
       %38 = triton_gpu.local_alloc %36 : (tensor<64x256xf8E5M2, #blocked1>) -> !tt.memdesc<64x256xf8E5M2, #shared1>
       // CHECK: triton_gpu.local_alloc
       // CHECK: scf.for
-      // CHECK:   tt.dot
-      // CHECK-NOT:  triton_nvidia_gpu.dot_async
+      // CHECK:   triton_nvidia_gpu.dot_async
+      // CHECK-NEXT: triton_nvidia_gpu.dot_wait
```

Comment:
This IR semantic is not fully correct. The IR looks like this now:
```
%a = triton_gpu.alloc %init -> !tt.memdesc...
%d = triton.gpu.dot_async %a, ...
%w = triton_nvidia_gpu.dot_wait %d
```
so it looks like the liverange %a dies when %d is created but it is actually used until %w.
That means when we assigned physical address to the shared memory we will consider that after inst defining %d the shared memory can be re-used which is incorrect. 
This could be solved in multiple ways but I think we want the semantic of a synchronous `dot` here.

Reply:
AIUI live range of the alloc is from the alloc until the dealloc, not having anything to do with the uses of the memdesc.  So I guess you are saying that dealloc can be hoisted above any instruction which does not explicitly use the memdesc?

If so then I think the problem is not that we need a sync-dot operation, but rather that dot_wait needs to use the memdesc that was passed to the dot.  

If dealloc can be hoisted as described and dot_wait does not use the memdesc, then we have the following problem.  Consider the following "normal pipelined loop".  This is the same before and after this patch.

```
memdesc = alloc();
for (i = 0; i < n; i++) {
  a = dot_async %memdesc
  dot_wait %a 1
}
dot_wait %a 0
dealloc(memdesc);
```

The final dot_wait does not mention the memdesc.  If it can be reordered with the dealloc, that is obviously wrong.

So one solution would be to explicitly mention the memdesc in the wait.  Another would be to have an alias analysis that "knows" which memdesc's the dot_wait uses.

I think having the AA makes more sense to me because mentioning the memdesc in the wait would be redundant.  We already have to mention the dot(s) in the wait, and from that you can trivially get all the memdesc's.  (In that respect this patch makes the IR more correct, because now we properly indicate which dots we're waiting on.)

Am I understanding this correctly or am I missing something?

Reply:
This seems weird to me.  One reason is that I would expect that unrolling a loop with known tripcount should always be a legal optimization, but consider:

```
%bufs = alloc space for 2 buffers
for i in range(2):
  %lhs = subview (i%2) %bufs
  dot_async %lhs
dot_wait 0
dealloc %bufs
```

becomes

```
%bufs = alloc space for 2 buffers
%lhs0 = subview 0 %bufs
dot_async %lhs0
%lhs1 = subview 1 %bufs
dot_async %lhs1
dot_wait 0
dealloc(%bufs)
```

But now either (a) the dealloc is allowed even though there's no loop, or (b) the dealloc is not allowed in which case loop unrolling has created illegal IR.

Another reason this feels weird to me is that we're saying that in the IR multi-buffered loops are a special case?  Normally we rely on live ranges to determine the lifetime of an alloc, but sometimes we don't.  Seems like we should do it one way or the other way, but not a mix?

Reply:
> This seems weird to me. One reason is that I would expect that unrolling a loop with known tripcount should always be a legal optimization, but consider:
> 
> ```
> %bufs = alloc space for 2 buffers
> for i in range(2):
>   %lhs = subview (i%2) %bufs
>   dot_async %lhs
> dot_wait 0
> dealloc %bufs
> ```
> 
> becomes
> 
> ```
> %bufs = alloc space for 2 buffers
> %lhs0 = subview 0 %bufs
> dot_async %lhs0
> %lhs1 = subview 1 %bufs
> dot_async %lhs1
> dot_wait 0
> dealloc(%bufs)
> ```
> 
> But now either (a) the dealloc is allowed even though there's no loop, or (b) the dealloc is not allowed in which case loop unrolling has created illegal IR.

dealloc is allowed even though there is no loop. What I meant to say is that dealloc can be used to explicitly set the liverange of an alloc (this is completely independent of loops or multi-buffering). If it is not set then the IR will assume the deallocation happens at the first common post-dominator point.

> 
> Also if this is legal IR, is it still legal IR if we break up the alloc into two separate allocs? (That can be a beneficial transformation because it gives us more flexibility in how we allocate the shmem.)
> 

Yes that's legal although I would assume it would be worse in term of performance in general

> Another reason this feels weird to me is that we're saying that in the IR multi-buffered loops are a special case? Normally we rely on live ranges to determine the lifetime of an alloc, but sometimes we don't. Seems like we should do it one way or the other way, but not a mix?

It's not a special case of multi-buffered, what I meant is that we currently don't set dealloc for other cases as it is not needed since the liverange can be correctly inferred from the IR. Always setting a dealloc would make it much harder to optimize the IR as we would constantly need to check for the dealloc and update the liverange.

Reply:
> Indeed based on the complexity of inserting and optimizing around deallocs, this should make our lives easier. For example we won't need an explicit dealloc at the end of the loop, because postdom analysis will Just Work correctly. (Possibly this would let us remove the dealloc operation entirely? I imagine the only reason it exists at all is to handle async ops?)

We actually do need the dealloc in this case (and that's why we added it in a first place) because the shared memory cannot be re-used right after the loop, it can only be used after the cp_asyn_wait 0 that ensures all the copies to the alloc are done. 

If the wait 0 outside had a reference to the alloc as you suggest then the dealloc could be removed indeed.

> We can make the async dot use the memdescs (a) explicitly, by making the memdescs operands of the wait, or (b) implicitly. It sounds like you prefer the explicit form so that there's no special case for dot_async, which I am happy to do.

Yes I think making memdescs an operand is the cleanest solution. I'm starting to think the wait should actually mirror exactly the signature of the dot as as all the resources used for the dot can actually be used asynchronously. (technically even register inputs cannot be re-used until the wait)

> I hear that you'd prefer to have synchronous dots. My concern about this as a solution is it creates a similar problem to the dealloc problem you noticed. If all of the dots are async, then we can have the invariant that an async dot is always consumed by a wait, and all uses of the dot's value (other than trivial ones like yield) go through that wait. This makes it easy to tell which waits "go with" which dots.

If I understand the problem correctly it is that:
```
%d0 = dot_async
%d1 = dot
```
in this case we have no wait so we don't have an explicit wait to feed %d0?

I agree adding explicit dependencies in the wait is the right solution and it seems to unify things.

Reply:
> If the wait 0 outside had a reference to the alloc as you suggest then the dealloc could be removed indeed.

Yeah, exactly.  Could be a nice starter project for someone, pytorch people were looking for projects?

I've gone and added the explicit operand dependencies.  PHAL!

Reply:
> > If the wait 0 outside had a reference to the alloc as you suggest then the dealloc could be removed indeed.
> 
> Yeah, exactly. Could be a nice starter project for someone, pytorch people were looking for projects?
> 
> I've gone and added the explicit operand dependencies. PHAL!

Sorry I realize now my comment was wrong, the dealloc is still needed to make sure the async_copies are done before we re-use the memory so the dealloc is not there only for async_mma and it is needed for mma_v2 anyway. 

So there is nothing to do, I think we want to keep what we have now